atomic_reactor/plugins/pre_reactor_config.py

```diff
@@ -0,0 +1,363 @@
+
+    def validate_required(self, conf):
+        """
+        Validate required paths.
+
+        :return: str list, errors
+        """
+        errors = []
+        for each_path in self.required:
+            errors.extend(self.validate_required_path(conf, required=each_path))
+
+        return errors
+        
+    def validate_required_path(self, conf, required, path=()):
+        """
+        Validate a single required path. Used by validate_required().
+
+        :return: str list, errors
+        """
+
+        if not required:
+            return []
+
+        errors = []
+        element = required[0]
+        if element is None:
+            # Check every key at this level
+            if conf is None:
+                errpath = '.'.join(path)
+                errors.append("{} must be map".format(errpath))
+                return errors
+
+            to_check = conf.keys()
+        elif isinstance(element, list):
+            # Check every value at this level
+            to_check = range(len(conf))
+        else:
+            to_check = [element]
+
+        for each_key in to_check:
+            subpath = path + (each_key,)
+            try:
+                subconf = conf[each_key]
+            except KeyError:
+                if len(required) > 1:
+                    # Not final element, ignore
+                    continue
+
+                errpath = '.'.join(subpath)
+                errors.append("{} missing".format(errpath))
+            except TypeError:
+                errpath = '.'.join(path)
+                errors.append("{} must be map".format(errpath))
+            else:
+                errs = self.validate_required_path(subconf,
+                                                   required=required[1:],
+                                                   path=subpath)
+                errors.extend(errs)
+
+        return errors
+
+
+class ClusterConfig(object):
+    """
+    Configuration relating to a particular cluster
+    """
+
+    def __init__(self, name, max_concurrent_builds=None, enabled=True):
+        self.name = str(name)
+        self.max_concurrent_builds = None
+        self.enabled = enabled
+
+        if max_concurrent_builds is not None:
+            self.max_concurrent_builds = int(max_concurrent_builds)
+
+
+class ReactorConfig(object):
+    """
+    Class to parse the atomic-reactor configuration file
+    """
+
+    CLUSTERS_KEY='clusters'
+    SCHEMA = {
+        'clusters': {
+            str: [
+                {
+                    'name': str,
+                    'max_concurrent_builds': int,
+                    'enabled': bool,
+                },
+            ]
+        }
+    }
+
+    REQUIRED = [
+        # clusters.*.name is required (but 'clusters' is not)
+        ('clusters', None, ['name']),
+    ]
+
+    def __init__(self, config_path=None, log=None):
+        self.log = log or logging.getLogger(__name__)
+
+        if config_path is None:
+            self.conf = {}
```

Comment:
I think for future proofing the code, it might be beneficial to also validate this default conf.